File: examples/tetris-duel/src/tetris_duel/board/pieces.py
"""The seven pieces, their rotations, and what a player calls each rotation."""

from __future__ import annotations

# Each piece as it looks on screen, read top row first.
SHAPES: dict[str, list[str]] = {
    "I": ["####"],
    "O": ["##", "##"],
    "T": ["###", ".#."],
    "S": [".##", "##."],
    "Z": ["##.", ".##"],
    "J": ["#..", "###"],
    "L": ["..#", "###"],
}


Cells = tuple[tuple[int, int], ...]


def _cells(rows: list[str]) -> Cells:
    return tuple((x, y) for y, row in enumerate(rows) for x, ch in enumerate(row) if ch == "#")


def _normalize(cells: Cells) -> Cells:
    left = min(x for x, _ in cells)
    top = min(y for _, y in cells)
    return tuple(sorted((x - left, y - top) for x, y in cells))


def _turn(cells: Cells) -> Cells:
    """One quarter turn clockwise."""
    bottom = max(y for _, y in cells)
    return _normalize(tuple((bottom - y, x) for x, y in cells))


def _orientations(rows: list[str]) -> tuple[Cells, ...]:
    """The distinct rotations of a piece: four for T/J/L, two for I/S/Z, one for O."""
    seen: list[Cells] = []
    shape = _normalize(_cells(rows))
    for _ in range(4):
        if shape not in seen:
            seen.append(shape)
        shape = _turn(shape)
    return tuple(seen)


PIECES: dict[str, tuple[Cells, ...]] = {name: _orientations(rows) for name, rows in SHAPES.items()}


ROTATION_NAMES = {
    "I": ("flat", "upright"),
    "O": ("square",),
    "T": ("pointing down", "pointing left", "pointing up", "pointing right"),
    "S": ("flat", "upright"),
    "Z": ("flat", "upright"),
    "J": ("hook left", "hook up", "hook right", "hook down"),
    "L": ("hook right", "hook down", "hook left", "hook up"),
}


def rotation_name(piece: str, rotation: int) -> str:
    names = ROTATION_NAMES.get(piece, ())
    return names[rotation] if rotation < len(names) else f"rotation {rotation}"

File: examples/tetris-duel/src/tetris_duel/static/js/game/pieces.js
/* The seven pieces and their rotations, as the server defines them. */
import { KEYS } from "./rules.js";
let shapes = {};
const rotations = {};
const normalize = (cells) => {
    const left = Math.min(...cells.map((c) => c[0]));
    const top = Math.min(...cells.map((c) => c[1]));
    return cells
        .map(([x, y]) => [x - left, y - top])
        .sort((a, b) => a[0] - b[0] || a[1] - b[1]);
};
const turn = (cells) => {
    const bottom = Math.max(...cells.map((c) => c[1]));
    return normalize(cells.map(([x, y]) => [bottom - y, x]));
};
/** Every distinct rotation of a piece drawn as rows of '#' and '.', in turning order. */
function orientations(rows) {
    const cells = [];
    rows.forEach((row, y) => [...row].forEach((ch, x) => ch === "#" && cells.push([x, y])));
    const out = [];
    let shape = normalize(cells);
    for (let i = 0; i < 4; i++) {
        const key = JSON.stringify(shape);
        if (!out.some((s) => JSON.stringify(s) === key))
            out.push(shape);
        shape = turn(shape);
    }
    return out;
}
/** Take the piece drawings from /api/setup, so the page and the server agree on every shape. */
export function loadPieces(drawings) {
    shapes = drawings;
    for (const key of KEYS)
        rotations[key] = orientations(drawings[key]);
}
export const drawing = (piece) => shapes[piece];
export const turns = (piece) => rotations[piece];
